src/app/TideSearchApplication.cpp

```diff
@@ -1675,87 +1741,67 @@ int TideSearchApplication::calcScoreCount(
+  double** dynProgArray = new double*[nCol];
+  for (col = 0; col < colFirst+maxDeltaMass; col++) {
+    dynProgArray[col] = new double[nRow];
+    memset(dynProgArray[col], 0.0, nRow*sizeof(double));
   }
-
-  dynProgArray[initCountRow][initCountCol] = 1.0; // initial count of peptides with mass = 1
+  
+  double* scoreCountBinAdjust = new double[nRow];
+  memset(scoreCountBinAdjust, 0.0, nRow*sizeof(double));
+  
+  // This part, below, is added by AKF. This must provide exactly the 
+  // same results as the original XPV implementation by Jeff Howbert, 
+  // albeit 2-3 times faster.
+  
+  dynProgArray[colStart][initCountRow] = 1.0; // initial count of peptides with mass = 1   // Modified by AKF
   vector<int> deltaMassCol(nDeltaMass);
   // populate matrix with scores for first (i.e. N-terminal) amino acid in sequence
   for (de = 0; de < nDeltaMass; de++) {
     ma = aaMass[de];
-    row = initCountRow + evidenceObs[ma + colStart];
-    col = initCountCol + ma;
+    col = colStart + ma;
+    row = initCountRow + evidenceObs[col];
     if (col <= maxDeltaMass + colLast) {
-      dynProgArray[row][col] += dynProgArray[initCountRow][initCountCol] * aaFreqN[de];
+      dynProgArray[col][row] += dynProgArray[colStart][initCountRow] * aaFreqN[de];
     }
   }
-  // set to zero now that score counts for first amino acid are in matrix
-  dynProgArray[initCountRow][initCountCol] = 0.0;
-  // populate matrix with score counts for non-terminal amino acids in sequence
+  for (ma = 0; ma < colFirst; ++ma )
+      delete dynProgArray[ma];
   for (ma = colFirst; ma < colLast; ma++) {
-    col = maxDeltaMass + ma;
-    evidence = evidenceObs[ma];
+    dynProgArray[ma+maxDeltaMass ] = new double[nRow];
+    memset(dynProgArray[ma+maxDeltaMass], 0.0, nRow*sizeof(double));
+    
     for (de = 0; de < nDeltaMass; de++) {
-      deltaMassCol[de] = col - aaMass[de];
+      deltaMassCol[de] = ma + aaMass[de];
     }
     for (row = rowFirst; row <= rowLast; row++) {
-      evidenceRow = row - evidence;
-      sumScore = dynProgArray[row][col];
-      for (de = 0; de < nDeltaMass; de++) {
-        sumScore += dynProgArray[evidenceRow][deltaMassCol[de]] * aaFreqI[de];
+      if (dynProgArray[ma][row] == 0.0) {
+        continue;
       }
-      dynProgArray[row][col] = sumScore;
-    }
-  }
-  // populate matrix with score counts for last (i.e. C-terminal) amino acid in sequence
-  ma = colLast;
-  col = maxDeltaMass + ma;
-  evidence = 0; // no evidence should be added for last amino acid in sequence
-  for (de = 0; de < nDeltaMass; de++) {
-    deltaMassCol[de] = col - aaMass[de];
-  }
-  for (row = rowFirst; row <= rowLast; row++) {
-    evidenceRow = row - evidence;
-    sumScore = 0.0;
-    for (de = 0; de < nDeltaMass; de++) {
-      sumScore += dynProgArray[evidenceRow][deltaMassCol[de]] * aaFreqC[de];  // C-terminal residue
+      for (de = 0; de < nDeltaMass; de++) {
+        col = deltaMassCol[de];
+        if (col == colLast) { 
```

Comment:
Wouldn't it be slightly faster to put the following two if statements into an if + else if?

if (newCol < colLast) { } else if (newCol == colLast)

Reply:
You are right,  I fixed it.